Kconfig.zephyr

```diff
@@ -494,6 +494,30 @@ config BUILD_OUTPUT_STRIPPED
 	  Build a stripped binary zephyr/zephyr.strip in the build directory.
 	  The name of this file can be customized with CONFIG_KERNEL_BIN_NAME.
 
+config BUILD_OUTPUT_ADJUST_LMA
+	string
+	help
+	  This will adjust the final ELF and hex output addresses LMA and VMA
+	  with the value provided.
+	  This will not affect the internal address symbols inside the image but
+	  can be useful when adjusting LMA addresses for flash tools or multi
+	  stage loaders where a pre-loader may copy image to a second location
+	  before booting second core.
```

Comment:
This is very contradictory, the config name implies only LMA is adjusted.

But the first sentence states that VMA is also adjusted.

And then the second sentence says that the "internal address symbols" are not affected. I assume "interal address symbols" is referring to VMA again. So this contradicts the first sentece.

Reply:
> This is very contradictory, the config name implies only LMA is adjusted.
>
> But the first sentence states that VMA is also adjusted.

That is because the purpose of this flag is to adjust the LMA so that the image may be automatically flashed at a different address or adjusted when converting to hex (bin files are unaffected by this as the don't contain any address information (LMA/VMA)).

When flashing an ELF, the flash tool will use the information provided by LMA to know where to flash the segments.

But when using `objcopy --change-addresses` then both LMA and VMA gets adjusted, not only the LMA.
Thus I find it important to make the user aware of this, although the purpose is to get LMA adjusted hence the reason for only LMA in the name (we don't care for the VMA in this particular case).

There could be other bintools, for example `elfconvert` with the ARM Compiler that supports adjusting only LMA, but so far I've focused only on GNU bintools.

> And then the second sentence says that the "internal address symbols" are not affected. I assume "interal address symbols" is referring to VMA again. So this contradicts the first sentece.

Internal address symbols are those used by the code and populated by the linker.
For example, if code refers to [__text_region_start](https://github.com/zephyrproject-rtos/zephyr/blob/5e05393afa5fa8ff291ad91428b8857f8ad47c1c/include/arch/arm/aarch32/cortex_m/scripts/linker.ld#L148) and that symbol points to address `0x00001212`, then the address of that symbol is unaltered when we change LMA. Meaning that if VMA is adjusted with `0x10000000`, then `__text_region_start` is still pointing to `0x00001212`.

I don't expect all developers to be aware of such details, thus I find it important to inform.

I will try to see if I can find a better description, but any proposals are more than welcome.
Hopefully you get the intention of the sentence now.

Reply:
> (bin files are unaffected by this as the don't contain any address information (LMA/VMA)).

VMA is encoded into every machine instruction generated so bin files do contain VMA information.

Reply:
> VMA is encoded into every machine instruction generated so bin files do contain VMA information.

correct, but it is not encoded as `VMA` information in itself.
In the binary it is purely an address, the VMA is a term and in the ELF file it's the `p_vaddr` when loading an ELF.
But bin files doesn't load an ELF when flashed (the native posix target does, cause it's using the ELF loader (`ld`)).
The bin file for embedded contains the information where to copy something as an address, but that's just it, an address.

So what is encoded as `p_vaddr` in the ELF and presented as VMA is lost as soon as the ELF is converted to binary.

Reply:
I see, you want to adjust the LMA, but are using a flag that adjusts both.

Could you try

--change-section-lma sectionpattern{=,+,-}val

with sectionpattern set to .*

Reply:
> --change-section-lma sectionpattern{=,+,-}val

I actually tried yesterday and failed to get desired behavior, but re-tried again now.

Happy to tell it works when pattern is correctly escaped:
```
--change-section-lma \*+<val>
```
will update PR, cause prefer to only adjust what strictly needs adjustment.

Reply:
Fixed, and VMA removed from the description which should make it clearer.